### PR TITLE
Add Pointer Authentication (ARMv8.3-PAuth) support to the TF

### PR DESCRIPTION
**Pull request by antonio-nino-diaz-arm**
This PR adds Pointer Authentication support to BL1, BL2 (in normal boot flow and BL2_AT_EL3), BL31 and the TSP.

It introduces two new build options:

- ``CTX_INCLUDE_PAUTH_REGS``: This will include all PAuth registers to the CPU context. If it isn't enabled, PAuth can't be used in the Secure world because the keys would be leaked to the Non-secure world. Disabling this option will cause accesses to PAuth registers from the Secure world to fault to prevent using them by mistake.

- ``ENABLE_PAUTH``: This enables PAuth in the images mentioned above, and requires ``CTX_INCLUDE_PAUTH_REGS=1``.

Note that the options have a significant impact in memory, and default to 0. There is some more information in the individual commit messages.

The only upstream platform that supports PAuth is FVP.

### CODE REVIEW

**Comment by antonio-nino-diaz-arm** [comment]
> func pauth_context_save
> ...
> mrs x9, APDBKeyLo_EL1
> mrs x10, APDBKeyHi_EL1
> stp x9, x10, [x11, #CTX_PACDAKEY_LO] -----> CTX_PACDBKEY_LO
> 
> func pauth_context_restore
> ...
> ldp x9, x10, [x11, #CTX_PACDAKEY_LO] -----> CTX_PACDBKEY_LO
> msr APDBKeyLo_EL1, x9
> msr APDBKeyHi_EL1, x10

Oh, thanks, I didn't see that.

**Comment by antonio-nino-diaz-arm** [comment]
The problem is that in order to remove that call, because of having asserts, you need a check such as
```
#if ENABLE_PAUTH || (CTX_INCLUDE_PAUTH_REGS && DEBUG)
```
in addition to all the checks inside ``bl_handle_pauth()``. I think that it becomes a bit of a maintainability issue then.

File: bl1/aarch64/bl1_entrypoint.S [hunk 1]
```diff
@@ -32,24 +32,42 @@ func bl1_entrypoint
...
-	 * platform memory map. Perform the latter here
-	 * and the former in bl1_main.
-	 * ---------------------------------------------
+	/* --------------------------------------------------------------------
+	 * Perform BL1 setup
+	 * --------------------------------------------------------------------
 	 */
-	bl	bl1_early_platform_setup
-	bl	bl1_plat_arch_setup
+	bl	bl1_setup
```
Comment (soby-mathew) [inline bl1/aarch64/bl1_entrypoint.S]: Better to unify the `bl1_setup()` sequence for AArch32 as well. The PAuth enable can then be invoked only for AArch64.

File: docs/user-guide.rst [hunk 2]
```diff
@@ -358,6 +358,12 @@ Common build options
    registers to be included when saving and restoring the CPU context. Default
    is 0.
 
+-  ``CTX_INCLUDE_PAUTH_REGS``: Boolean option that, when set to 1, will cause
+   the ARMv8.3-PAuth registers to be included when saving and restoring the CPU
+   context. Note that if the hardware supports this extension and this option is
+   set to 0 the value of the registers will be leaked between Secure and
+   Non-secure worlds. The default is 0.
```
[the remark this answers is not in the document]
Comment (antonio-nino-diaz-arm) [reply docs/user-guide.rst]: Done

File: docs/firmware-design.rst [hunk 3]
```diff
@@ -2566,6 +2566,11 @@ Armv8.3-A
    must be set to 1. This will add all pointer authentication system registers
    to the context that is saved when doing a world switch.
 
+   The Trusted Firmware itself has support for pointer authentication at runtime
+   that can be enabled by setting both options ``ENABLE_PAUTH`` and
+   ``CTX_INCLUDE_PAUTH_REGS`` to 1. This enables pointer authentication in BL1,
+   BL2, BL31, and the TSP if it is used.
+
```
[the remark this answers is not in the document]
Comment (antonio-nino-diaz-arm) [reply docs/firmware-design.rst]: This will be done as part of a future patch.

File: bl1/aarch64/bl1_entrypoint.S (hunk 1 above)
Comment (antonio-nino-diaz-arm) [reply bl1/aarch64/bl1_entrypoint.S]: Done

File: bl1/bl1_main.c [hunk 4]
```diff
@@ -51,6 +51,28 @@ void bl1_calc_bl2_mem_layout(const meminfo_t *bl1_mem_layout,
...
 
+/*******************************************************************************
+ * Setup function for BL1.
+ ******************************************************************************/
+void bl1_setup(void)
+{
+	/* Perform early platform-specific setup */
+	bl1_early_platform_setup();
+
+#if AARCH64
```
Comment (soby-mathew) [inline bl1/bl1_main.c]: use #ifdef here.

Comment (antonio-nino-diaz-arm) [reply bl1/bl1_main.c]: Done. Fixed the other occurrences as well.

**Pull request merged**